addons/account/models/account_payment.py

```diff
@@ -457,6 +473,190 @@ def unlink(self):
+                'ref': payment.communication,
+                'journal_id': payment.journal_id.id,
+                'currency_id': payment.journal_id.currency_id.id or payment.company_id.currency_id.id,
+                'partner_id': payment.partner_id.id,
+                'line_ids': [
+                    # Receivable / Payable / Transfer line.
+                    (0, 0, {
+                        'name': rec_pay_line_name,
+                        'amount_currency': counterpart_amount + write_off_amount,
+                        'currency_id': currency_id,
+                        'debit': balance + write_off_balance > 0.0 and balance + write_off_balance or 0.0,
+                        'credit': balance + write_off_balance < 0.0 and -balance - write_off_balance or 0.0,
+                        'date_maturity': payment.payment_date,
+                        'partner_id': payment.partner_id.id,
+                        'account_id': payment.destination_account_id.id,
+                        'payment_id': payment.id,
+                    }),
+                    # Liquidity line.
+                    (0, 0, {
+                        'name': liquidity_line_name,
+                        'amount_currency': -liquidity_amount,
+                        'currency_id': liquidity_line_currency_id,
+                        'debit': balance < 0.0 and -balance or 0.0,
+                        'credit': balance > 0.0 and balance or 0.0,
+                        'date_maturity': payment.payment_date,
+                        'partner_id': payment.partner_id.id,
+                        'account_id': liquidity_line_account.id,
+                        'payment_id': payment.id,
+                    }),
+                ],
+            }
+            if write_off_balance:
+                # Write-off line.
+                move_vals['line_ids'].append((0, 0, {
+                    'name': payment.writeoff_label,
+                    'amount_currency': -write_off_amount,
+                    'currency_id': currency_id,
+                    'debit': write_off_balance < 0.0 and -write_off_balance or 0.0,
+                    'credit': write_off_balance > 0.0 and write_off_balance or 0.0,
+                    'date_maturity': payment.payment_date,
+                    'partner_id': payment.partner_id.id,
+                    'account_id': payment.writeoff_account_id.id,
+                    'payment_id': payment.id,
+                }))
+
+            all_move_vals.append(move_vals)
+
+            # ==== 'transfer' ====
+            if payment.payment_type == 'transfer':
+
+                if payment.destination_journal_id.currency_id:
+                    transfer_amount = payment.currency_id._convert(
+                        counterpart_amount, payment.destination_journal_id.currency_id, payment.company_id, payment.payment_date)
+                else:
+                    transfer_amount = 0.0
+
+                move_vals = {
```

Comment:
reusing move_vals is a bit confusing (especially when you consider the length of this function), maybe just call it transfer_move_vals